pygaggle/data/msmarco.py

```diff
@@ -0,0 +1,145 @@
+                    print(f'Loading queries {i}')
+        return queries
+
+    @classmethod
+    def from_folder(cls, 
+                    folder: str, 
+                    split: str = 'dev', 
+                    is_duo: bool = False) -> 'MsMarcoDataset':
+        run_mono = "mono." if is_duo else ""
+        query_path = os.path.join(folder, f"queries.{split}.small.tsv")
+        qrels_path = os.path.join(folder, f"qrels.{split}.small.tsv")
+        run_path = os.path.join(folder, f"run.{run_mono}{split}.small.tsv")
+        return cls(examples = cls.load_queries(query_path, 
+                                               cls.load_qrels(qrels_path),
+                                               cls.load_run(run_path)))
+
+
+    def query_passage_tuples(self, is_duo: bool = False):
+        return (((ex.qid, ex.text, ex.relevant_candidates), perm_pas) for ex in self.examples
+                for perm_pas in permutations(ex.candidates, r=1+int(is_duo)))
+
+
+    def to_relevance_examples(self,
+                              index_path: str,
+                              is_duo: bool = False) -> List[RelevanceExample]:
+        loader = MsMarcoPassageLoader(index_path)
+        example_map = {}
+        for (qid, text, rel_cands), cands in self.query_passage_tuples():
+            if qid not in example_map:
+                example_map[qid] = [convert_to_unicode(text), [], [], []]
+            #TODO generalize for duoBERT by removing indexing to 0
+            example_map[qid][1].append([cand for cand in cands][0])
+            try:
+                passages = [loader.load_passage(cand) for cand in cands]
+                #TODO generalize for duoBERT by removing indexing to 0
+                example_map[qid][2].append([convert_to_unicode(passage.all_text) for passage in passages][0]) 
+            except ValueError as e:
+                logging.warning(f'Skipping {passages}')
+                continue
+            example_map[qid][3].append(cands[0] in rel_cands)
+        mean_stats = defaultdict(list)
+        for ex in self.examples:
+            int_rels = np.array(list(map(int, example_map[ex.qid][3])))
+            p = int_rels.sum()/(len(ex.candidates) - 1) if is_duo else int_rels.sum()
+            mean_stats['Random P@1'].append(np.mean(int_rels))
+            n = len(ex.candidates) - p
+            N = len(ex.candidates)
+            if len(ex.candidates) <= 1000:
+                mean_stats['Random R@1000'].append(1 if 1 in int_rels else 0)
+            numer = np.array([sp.comb(n, i) / (N - i) for i in range(0, n + 1) if i!=N]) * p
+            if n == N:
+                numer = np.append(numer, 0)
+            denom = np.array([sp.comb(N, i) for i in range(0, n + 1)])
+            rr = 1 / np.arange(1, n + 2)
+            rmrr = np.sum(numer * rr / denom)
+            mean_stats['Random MRR'].append(rmrr)
+            rmrr10 = np.sum(numer[:10] * rr[:10] / denom[:10]) #TODO verify if this works
```

Comment:
Should be right.

Reply:
Yep! Forgot to uncomment some of these TODOs :)